pkg/apis/pipeline/v1/matrix_types.go

```diff
@@ -96,11 +97,30 @@ func createCombination(name string, value string, combination Params) Params {
 
 // CountCombinations returns the count of combinations of Parameters generated from the Matrix in PipelineTask.
 func (m *Matrix) CountCombinations() int {
```

Comment:
this function has kind of three separate code paths:

1. non-zero matrix.Include and no matrix.Params
2. non-zero matrix.Params
3. non-zero matrix.Include

we could simplify this a little further by creating a separate function to count matrix.Params combinations and then overwriting or updating the count based on matrix.Include.

Reply:
something like:

```
func (m *Matrix) countParamCombinations() int {
	count := 1
	// Iterate over matrix.params and compute count of all generated combinations
	for _, param := range m.Params {
		count *= len(param.Value.ArrayVal)
	}

	return count
}

// CountCombinations returns the count of combinations of Parameters generated from the Matrix in PipelineTask.
func (m *Matrix) CountCombinations() int {
	// Explicit combinations in the Matrix
	// Calculate based on length of include Names and return count
	count := m.countParamCombinations()
	if count == 0 {
		return len(m.Include)
	}
	paramMap := m.Params.extractParamMapArrVals()
	// Check which combinations will be replaced vs generated based on matrix include params
	for _, include := range m.Include {
		for _, param := range include.Params {
			if val, exist := paramMap[param.Name]; exist {
				// If the matrix include param values does not exist, a new combination will be generated
				if !slices.Contains(val, param.Value.StringVal) {
					count++
				} else {
					break
				}
			}
		}
	}
	return count
}
```

Reply:
Thanks @pritidesai. I have extracted into  countMatrixParamsCombinations() and countNewCombinationsFromInclude()